ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -467,25 +418,143 @@ private void testVariousErrors(String dcWherePutHappened)
     getAndAssertSuccess();
   }
 
-  // @todo: test the case where read is not called before chunks come in.
-  // @todo: test the case where read comes in immediately after callback, and chunks come in delayed.
-  // @todo: test the case where operation callback is called with exception.
-  // @todo: test the case where async write results in an exception. - read should be notified,
-  //        operation should get completed.
-  // @todo: test the case where a subsequent chunk fails at the router. the read should fail in that case.
-  // @todo: maybe a test for legacy blobs.
-  // @todo: possibly tests where intermediate chunks get expired/deleted.
+  /**
+   * Test that read succeeds when all chunks are received before read is called.
+   * @throws Exception
+   */
+  @Test
+  public void testReadNotCalledBeforeChunkArrival()
+      throws Exception {
+    // 3 chunks so blob can be cached completely before reading
+    blobSize = maxChunkSize * 2 + 1;
+    doPut();
+    getAndAssertSuccess(true, null);
+  }
+
+  /**
+   * Test that read succeeds when read is called immediately after callback, and chunks come in delayed.
+   * @throws Exception
+   */
+  @Test
+  public void testDelayedChunks()
+      throws Exception {
+    doPut();
+    final CountDownLatch readIntoLatch = new CountDownLatch(1);
+    RouterTestHelpers.setDataBlobGetLatchForAllServers(readIntoLatch, mockServerLayout);
+    getAndAssertSuccess(false, readIntoLatch);
+    RouterTestHelpers.setDataBlobGetLatchForAllServers(null, mockServerLayout);
+  }
+
+  /**
+   * Test that data chunk errors notify the reader callback and set the error code correctly.
+   * @throws Exception
+   */
+  @Test
+  public void testDataChunkFailure()
+      throws Exception {
+    for (ServerErrorCode serverErrorCode : ServerErrorCode.values()) {
+      if (serverErrorCode != ServerErrorCode.No_Error) {
+        testDataChunkError(serverErrorCode, RouterErrorCode.UnexpectedInternalError);
+      }
+    }
+  }
+
+  /**
+   * Test that gets work for blobs with the old blob format (V1).
+   * @throws Exception
+   */
+  @Test
+  public void testLegacyBlobGetSuccess()
+      throws Exception {
+    RouterTestHelpers.setBlobFormatForAllServers(false, mockServerLayout);
+    for (int i = 0; i < 10; i++) {
+      // blobSize in the range [1, maxChunkSize]
+      blobSize = random.nextInt(maxChunkSize) + 1;
+      doPut();
+      getAndAssertSuccess();
+    }
+    RouterTestHelpers.setBlobFormatForAllServers(true, mockServerLayout);
+  }
+
+  // @todo: test operation completion behavior when stream is not read completely. This will require addin
```

Comment:
Yep, we discussed this and decided on adding the timeout behavior in the future.  I'll take out this todo message, though.